storm-core/src/jvm/backtype/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -28,94 +32,270 @@
-                                    td.getName(), execsNeedScheduling, targetNode.getHostname(), targetSlot.getPort());
-                            if (!nodesUsed.contains(targetNode.getId())) {
-                                nodesUsed.add(targetNode.getId());
+            if (schedulingPrioritystrategy == null) {
+                try {
+                    schedulingPrioritystrategy = (ISchedulingPriorityStrategy) Utils.newInstance((String) this.conf.get(Config.RESOURCE_AWARE_SCHEDULER_PRIORITY_STRATEGY));
+                } catch (RuntimeException e) {
+                    LOG.error("failed to create instance of priority strategy: {} with error: {}! No topologies will be scheduled.",
+                            this.conf.get(Config.RESOURCE_AWARE_SCHEDULER_PRIORITY_STRATEGY), e.getMessage());
+                    break;
+                }
+            }
+            TopologyDetails td = null;
+            try {
+                //need to re prepare since scheduling state might have been restored
+                schedulingPrioritystrategy.prepare(this.topologies, this.cluster, this.userMap, this.nodes);
+                //Call scheduling priority strategy
+                td = schedulingPrioritystrategy.getNextTopologyToSchedule();
+            } catch (Exception e) {
+                LOG.error("Exception thrown when running priority strategy {}. No topologies will be scheduled! Error: {} StackTrace: {}"
+                        , schedulingPrioritystrategy.getClass().getName(), e.getMessage(), Arrays.toString(e.getStackTrace()));
+                break;
+            }
+            if (td == null) {
+                break;
+            }
+            scheduleTopology(td);
+        }
+    }
+
+    public void scheduleTopology(TopologyDetails td) {
+        User topologySubmitter = this.userMap.get(td.getTopologySubmitter());
+        if (cluster.getUnassignedExecutors(td).size() > 0) {
+            LOG.debug("/********Scheduling topology {} from User {}************/", td.getName(), topologySubmitter);
+
+            SchedulingState schedulingState = this.checkpointSchedulingState();
+            IStrategy rasStrategy = null;
+            try {
+                rasStrategy = (IStrategy) Utils.newInstance((String) td.getConf().get(Config.TOPOLOGY_SCHEDULER_STRATEGY));
+            } catch (RuntimeException e) {
+                LOG.error("failed to create instance of IStrategy: {} with error: {}! Topology {} will not be scheduled.",
+                        td.getName(), td.getConf().get(Config.TOPOLOGY_SCHEDULER_STRATEGY), e.getMessage());
+                this.restoreCheckpointSchedulingState(schedulingState);
+                //since state is restored need the update User topologySubmitter to the new User object in userMap
+                topologySubmitter = this.userMap.get(td.getTopologySubmitter());
+                topologySubmitter.moveTopoFromPendingToInvalid(td);
```

Comment:
```
this.restoreCheckpointSchedulingState(schedulingState);
//since state is restored need the update User topologySubmitter to the new User object in userMap
topologySubmitter = this.userMap.get(td.getTopologySubmitter());
```

Yeah, I was focusing on these lines.  I think I clicked the wrong hover-over plus button.  Even the comment line appears verbatim many places.